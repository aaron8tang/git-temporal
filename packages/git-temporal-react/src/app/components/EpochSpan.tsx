import React from 'react';
import { style } from 'app/styles';

export interface EpochSpanProps {
  firstEpochTime: number;
  secondEpochTime: number;
  style?: object | string;
}

const secondsPer = {
  year: 31556926,
  month: 2629743,
  week: 604800,
  day: 86400,
  hour: 3600,
  minute: 60,
};

export function convertEpochDateTimesToTimeSpan(
  firstEpochTime: number,
  secondEpochTime: number
): string {
  const timeSpan = Math.abs(firstEpochTime - secondEpochTime);
  for (const timeComponent in secondsPer) {
    const seconds = secondsPer[timeComponent];
    if (timeSpan > seconds) {
      const roundedNumberForComponent = Math.round(timeSpan / seconds);
      const label =
        roundedNumberForComponent === 1 ? timeComponent : `${timeComponent}s`;
      return `${roundedNumberForComponent} ${label}`;
    }
  }
  return '0 seconds';
}
export const EpochSpan = (props: EpochSpanProps): JSX.Element => {
  return (
    <span style={style(props.style)}>
      {convertEpochDateTimesToTimeSpan(
        props.firstEpochTime,
        props.secondEpochTime
      )}
    </span>
  );
};

EpochSpan.displayName = 'EpochSpan';
