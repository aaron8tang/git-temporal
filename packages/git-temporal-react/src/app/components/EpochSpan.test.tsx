import React from 'react';
import { shallow } from 'enzyme';

import { EpochSpan } from './EpochSpan';

const firstEpochTime = 1536691136;
const secondEpochTime = 1538704960;

describe('components/EpochSpan', () => {
  describe('when rendered with times in order', () => {
    let wrapper;
    beforeAll(() => {
      wrapper = shallow(
        <EpochSpan
          firstEpochTime={firstEpochTime}
          secondEpochTime={secondEpochTime}
        />
      );
    });

    test('it should match snapshot (it should have rendered "3 weeks")', () => {
      expect(wrapper).toMatchSnapshot();
    });
  });
  describe('when rendered with times in reverse order', () => {
    let wrapper;
    beforeAll(() => {
      wrapper = shallow(
        <EpochSpan
          firstEpochTime={secondEpochTime}
          secondEpochTime={firstEpochTime}
        />
      );
    });

    test('it should match snapshot (it should still have rendered "3 weeks")', () => {
      expect(wrapper).toMatchSnapshot();
    });
  });
});
