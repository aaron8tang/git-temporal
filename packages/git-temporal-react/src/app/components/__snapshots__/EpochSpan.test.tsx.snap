// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/EpochSpan when rendered with times in order it should match snapshot (it should have rendered "3 weeks") 1`] = `
<span
  style={Object {}}
>
  3 weeks
</span>
`;

exports[`components/EpochSpan when rendered with times in reverse order it should match snapshot (it should still have rendered "3 weeks") 1`] = `
<span
  style={Object {}}
>
  3 weeks
</span>
`;
